linkerd/app/outbound/src/http/logical/policy/route/filters.rs

```diff
@@ -0,0 +1,197 @@
+        pub location: ::http::Uri,
+    }
+
+    #[derive(Debug, thiserror::Error, Clone, PartialEq, Eq)]
+    #[error("HTTP request configured to fail with {status}: {message}")]
+    pub struct HttpRouteInjectedFailure {
+        pub status: ::http::StatusCode,
+        pub message: Arc<str>,
+    }
+
+    #[derive(Debug, thiserror::Error, Clone, PartialEq, Eq)]
+    #[error("gRPC request configured to fail with {code}: {message}")]
+    pub struct GrpcRouteInjectedFailure {
+        pub code: u16,
+        pub message: Arc<str>,
+    }
+
+    #[derive(Debug, thiserror::Error)]
+    #[error("invalid server policy: {0}")]
+    pub struct HttpInvalidPolicy(pub &'static str);
+}
+
+pub(crate) trait Apply {
+    fn apply<B>(&self, req: &mut ::http::Request<B>) -> Result<()>;
+}
+
+pub fn apply_http<B>(
+    r#match: &http::RouteMatch,
+    filters: &[http::Filter],
+    req: &mut ::http::Request<B>,
+) -> Result<()> {
+    // TODO Do any metrics apply here?
+    for filter in filters {
+        match filter {
+            policy::http::Filter::InjectFailure(fail) => {
+                if let Some(policy::http::filter::FailureResponse { status, message }) =
+                    fail.apply()
+                {
+                    return Err(errors::HttpRouteInjectedFailure { status, message }.into());
+                }
+            }
+
+            policy::http::Filter::Redirect(redir) => match redir.apply(req.uri(), r#match) {
+                Ok(Some(policy::http::filter::Redirection { status, location })) => {
+                    return Err(errors::HttpRouteRedirect { status, location }.into());
+                }
+
+                Err(invalid) => {
+                    return Err(errors::HttpRouteInvalidRedirect(invalid).into());
+                }
+
+                Ok(None) => {
+                    tracing::debug!("Ignoring irrelevant redirect");
+                }
+            },
+
+            policy::http::Filter::RequestHeaders(rh) => {
+                rh.apply(req.headers_mut());
+            }
+
+            policy::http::Filter::InternalError(msg) => {
+                return Err(errors::HttpInvalidPolicy(msg).into());
+            }
+        }
+    }
+
+    Ok(())
+}
+
+pub fn apply_grpc<B>(
+    _match: &grpc::RouteMatch,
+    filters: &[grpc::Filter],
+    req: &mut ::http::Request<B>,
+) -> Result<()> {
+    for filter in filters {
+        match filter {
+            policy::grpc::Filter::InjectFailure(fail) => {
+                if let Some(policy::grpc::filter::FailureResponse { code, message }) = fail.apply()
+                {
+                    return Err(errors::GrpcRouteInjectedFailure { code, message }.into());
+                }
+            }
+
+            policy::grpc::Filter::RequestHeaders(rh) => {
+                rh.apply(req.headers_mut());
+            }
+
+            policy::grpc::Filter::InternalError(msg) => {
+                return Err(errors::HttpInvalidPolicy(msg).into());
+            }
```

Comment:
very tiny nit: `policy::grpc` is already imported in this module:
```suggestion
           grpc::Filter::InjectFailure(fail) => {
                if let Some(grpc::filter::FailureResponse { code, message }) = fail.apply()
                {
                    return Err(errors::GrpcRouteInjectedFailure { code, message }.into());
                }
            }

            grpc::Filter::RequestHeaders(rh) => {
                rh.apply(req.headers_mut());
            }

            grpc::Filter::InternalError(msg) => {
                return Err(errors::HttpInvalidPolicy(msg).into());
            }
```